pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -0,0 +1,552 @@
+		klog.V(2).InfoS("No Gateway is found", "err", err)
+		return nil
+	}
+	activeGW := c.getActiveGateway(gws)
+	ciImp, err := c.ciImportLister.ClusterInfoImports(c.namespace).Get(k.name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			cachedCI, isExist, _ := c.installedCIImports.GetByKey(k.name)
+			if isExist {
+				cachedata := cachedCI.(*ciImportInfo)
+				klog.InfoS("Deleting remote Node flows for multi-cluster traffic", "kind", k.kind, "name", k.name)
+				return c.deleteMCNodeFlowsForCIImport(&cachedata.ciImport,
+					getCacheKeyWithPrefix(activeGW.Name+cachedata.ciImport.Spec.ClusterID))
+			}
+			klog.V(2).InfoS("No cached ClusterInfoImport found, do nothing")
+			return nil
+		}
+		return err
+	}
+	klog.V(2).InfoS("Add multic-cluster flows for ClusterInfoImport", "clusterinfoimport", ciImp.Name)
+	if err = c.addMCNodeFlows(activeGW, ciImp); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *MCRouteController) addMCNodeFlows(gw *mcv1alpha1.Gateway, ciImp *mcv1alpha1.ClusterInfoImport) error {
+	allciImports, err := c.getAllCIImports(ciImp)
+	if allciImports == nil {
+		return err
+	}
+	var tunnelPeerIP net.IP
+	for _, ciImport := range allciImports {
+		tunnelPeerIP = getPeerGatewayIP(ciImport.Spec)
+		if tunnelPeerIP == nil {
+			klog.V(2).InfoS("No valid peer Gateway IP from ClusterInfoImport, skip updating openflow rules",
+				"clusterinfoimport", klog.KObj(ciImport), "node", c.nodeConfig.Name)
+			continue
+		}
+
+		klog.InfoS("Adding/Updating remote Gateway Node flows for Multi-cluster", "gateway", klog.KObj(gw),
+			"node", c.nodeConfig.Name, "peer", tunnelPeerIP)
+		allCIDRs := []string{ciImport.Spec.ServiceCIDR}
+		peerConfigs, err := generatePeerConfigs(allCIDRs, tunnelPeerIP)
+		if err != nil {
+			klog.ErrorS(err, "Parse error for serviceCIDR from remote cluster", "clusterinfoimport", ciImport.Name, "gateway", gw.Name)
+			return err
+		}
+		if c.iAmActiveGW {
+			klog.V(2).InfoS("Adding/Updating flows to remote Gateway Node for Multi-cluster traffic", "clusterinfoimport", ciImport.Name, "cidrs", allCIDRs)
+			localGatewayIP := net.ParseIP(gw.GatewayIP)
+			if err := c.ofClient.InstallMulticlusterGatewayFlows(
+				getCacheKeyWithPrefix(gw.Name+ciImport.Spec.ClusterID),
+				peerConfigs,
+				tunnelPeerIP,
+				localGatewayIP); err != nil {
+				return fmt.Errorf("failed to install flows to remote Gateway in ClusterInfoImport %s: %v", ciImport.Name, err)
+			}
+		} else {
+			klog.V(2).InfoS("Adding/Updating flows to the local active Gateway for Multi-cluster traffic", "clusterinfoimport", ciImport.Name, "cidrs", allCIDRs)
+			tunnelPeerIP := net.ParseIP(gw.InternalIP)
+			if err := c.ofClient.InstallMulticlusterNodeFlows(
+				getCacheKeyWithPrefix(gw.Name+ciImport.Spec.ClusterID),
+				peerConfigs,
+				tunnelPeerIP); err != nil {
+				return fmt.Errorf("failed to install flows to Gateway %s: %v", gw.Name, err)
+			}
+		}
+
+		c.installedCIImports.Add(&ciImportInfo{
+			localGWName: gw.Name,
```

Comment:
Do we need to save the GW name for each CIImport, if we already save installedGW?